source/fab/errors.py

```diff
@@ -0,0 +1,344 @@
+    ) -> None:
+        self.category = category
+        self.expected = expected
+
+        super().__init__(tool, f"got type {category} instead of {expected}")
+
+
+class FabToolInvalidVersion(FabToolError):
+    """Version format problem.
+
+    Error when version information cannot be extracted from a specific
+    tool.  Where a version pattern is available, report this as part
+    of the error.
+
+    :param tool: name of the current tool
+    :param value: output from the query command
+    :param expected: optional format of version string
+    """
+
+    def __init__(
+        self,
+        tool: Union["Category", "Tool", str],
+        value: str,
+        expected: Optional[str] = None,
+    ) -> None:
+        self.value = value
+        self.expected = expected
+
+        message = f"invalid version {repr(self.value)}"
+        if expected is not None:
+            message += f" should be {repr(expected)}"
+
+        super().__init__(tool, message)
+
+
+class FabToolPsycloneAPI(FabToolError):
+    """PSyclone API and target problem.
+
+    Error when the specified PSyclone API, which can be empty to
+    indicate DSL mode, and the associated target do not match.
+
+    :param api: the current API or empty for DSL mode
+    :param target: the name of the target
+    :param present: optionally whether the target is present or
+        absent.  Used to format the error message.  Defaults to False.
+    """
+
+    def __init__(
+        self, api: Union[str, None], target: str, present: Optional[bool] = False
+    ) -> None:
+        self.target = target
+        self.present = present
+        self.api = api
+
+        message = "called "
+        if api:
+            message += f"with {api} API "
+        else:
+            message += "without API "
+        if present:
+            message += "and with "
+        else:
+            message += "but not with "
+        message += f"{target}"
+
+        super().__init__("psyclone", message)
+
+
+class FabToolNotAvailable(FabToolError):
+    """An unavailable tool has been requested.
+
+    Error where a tool which is not available in a particular suite of
+    tools has been requested.
+
+    :param tool: name of the current tool
+    :param suite: optional name of the current tool suite
+    """
+
+    def __init__(
+        self, tool: Union["Category", "Tool", str], suite: Optional[str] = None
+    ) -> None:
+        message = "not available"
+        if suite:
+            message += f" in suite {suite}"
+        super().__init__(tool, message)
+
+
+class FabToolInvalidSetting(FabToolError):
+    """An invalid tool setting has been requested.
+
+    Error where an invalid setting, e.g. MPI, has been requested for a
+    particular tool.
+
+    :param setting_type: name of the invalid setting
+    :param tool: the tool to which setting applies
+    :param additional: optional additional information
+    """
+
+    # FIXME: improve these here and in tool_repository
```

Comment:
Ideally we will eventually remove the need for this failure mode in that the API will not allow impossible requests to be made. Clearly not an issue for this change.